components/org.wso2.carbon.identity.application.authenticator.adapter/src/main/java/org/wso2/carbon/identity/application/authenticator/adapter/internal/AuthenticationRequestBuilder.java

```diff
@@ -110,23 +114,32 @@ private User getUserForEventBuilder(AuthenticatedUser authenticatedUser)
         }
     }
 
-    private Organization getOrganizationForEventBuilder(AuthenticatedUser authenticatedUser) {
+    private void setTenantAndOrganizationToEventBuilder(AuthenticationRequestEvent.Builder eventBuilder,
+                                                        AuthenticationContext context)
+            throws ActionExecutionRequestBuilderException {
+
+        String tenantDomain = context.getTenantDomain();
 
-        String organizationId = authenticatedUser.getUserResidentOrganization();
         try {
-            if (organizationId != null && !organizationId.isEmpty()) {
-                String organizationName = AuthenticatorAdapterDataHolder.getInstance().getOrganizationManager()
-                        .getOrganizationNameById(authenticatedUser.getUserResidentOrganization());
-                return new Organization(authenticatedUser.getUserResidentOrganization(), organizationName);
+            /* Only if the user is attempting to log in to a sub-organization, the organization will be set,
+             and also the root tenant will be considered the tenant domain. */
+            if (OrganizationManagementUtil.isOrganization(tenantDomain)) {
+                eventBuilder.organization(new Organization(
+                        tenantDomain,
+                        organizationManager.getOrganizationNameById(tenantDomain)
+                ));
+                String primaryOrganizationId = organizationManager.getPrimaryOrganizationId(tenantDomain);
+                tenantDomain = organizationManager.getOrganizationNameById(primaryOrganizationId);
```

Comment:
Will this give the root or the immediate parent when there's a hierarchy

Reply:
This gives the root tenant.